File: 32_02_PrintTreesInLines/PrintTreesInLines.cpp
#include <iostream>
#include <vector>
#include <string>
#include <deque>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <algorithm>
#include <cstring>
using namespace std;

struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution
{
  public:
    void PrintInLines(TreeNode *root)
    {
        queue<TreeNode *> q;
        if (root == nullptr)
            return;
        q.push(root);
        int nextLevel = 0;
        int toBePrinted = 1;
        while (!q.empty())
        {
            root = q.front();
            printf("%d ", root->val);
            if (root->left)
            {
                q.push(root->left);
                ++nextLevel;
            }
            if (root->right)
            {
                q.push(root->right);
                ++nextLevel;
            }
            q.pop();
            --toBePrinted;
            if (toBePrinted == 0)
            {
                printf("\n");
                toBePrinted = nextLevel;
                nextLevel = 0;
            }
        }
    }
};

class Solution1
{
  public:
    vector<vector<int>> Print(TreeNode *pRoot)
    {
        vector<vector<int>> ans;
        if (pRoot == nullptr)
            return ans;
        queue<TreeNode *> q;
        q.push(pRoot);
        while (!q.empty())
        {
            int size = q.size();
            vector<int> levelElem;
            while (size--)
            {
                TreeNode *t = q.front();
                q.pop();
                levelElem.push_back(t->val);
                if (t->left)
                    q.push(t->left);
                if (t->right)
                    q.push(t->right);
            }
            ans.push_back(levelElem);
        }
        return ans;
    }
};

File: 32_02_PrintTreesInLines/PrintTreesInLines.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def PrintInLines(self, root):
        if not root:
            return
        queue = [root]
        nextLevel = 0
        toBePrinted = 1
        while queue:
            root = queue.pop(0)
            print("%d " % root.val, end='')
            if root.left:
                queue.append(root.left)
                nextLevel += 1
            if root.right:
                queue.append(root.right)
                nextLevel += 1
            toBePrinted -= 1
            if toBePrinted == 0:
                print()
                toBePrinted = nextLevel
                nextLevel = 0


class Solution1:
    def Print(self, pRoot):
        # write code here
        if not pRoot:
            return []
        L = []
        queue = [pRoot]
        while queue:
            levelElem = []
            for i in queue:
                levelElem.append(i.val)
            L.append(levelElem)
            for i in range(len(queue)):
                root = queue.pop(0)
                if root.left:
                    queue.append(root.left)
                if root.right:
                    queue.append(root.right)
        return L


t1 = TreeNode(8)
t2 = TreeNode(6)
t3 = TreeNode(10)
t1.left = t2
t1.right = t3
s = Solution1()
print(s.Print(t1))
